compiler/rustc_middle/src/ty/layout.rs

```diff
@@ -1078,6 +1076,34 @@ impl<'tcx> LayoutCx<'tcx, TyCtxt<'tcx>> {
 
                     let mut st = self.univariant_uninterned(ty, &variants[v], &def.repr(), kind)?;
                     st.variants = Variants::Single { index: v };
+
+                    if def.is_unsafe_cell() {
+                        let fill = |scalar: &mut _| match scalar {
+                            Scalar::Initialized { value, valid_range } => {
+                                *valid_range = WrappingRange::full(value.size(dl))
+                            }
+                            // Already doesn't have any niches
+                            Scalar::Union { .. } => {}
+                        };
+                        match &mut st.abi {
+                            Abi::Uninhabited => {}
+                            Abi::Scalar(scalar) => fill(scalar),
+                            Abi::ScalarPair(a, b) => {
+                                fill(a);
+                                fill(b);
+                            }
+                            Abi::Vector { element, count: _ } => {
+                                // Until we support types other than floats and integers in SIMD,
+                                // `element` must already be a full for its range, so there's nothing to
+                                // do here.
+                                assert!(element.is_always_valid(dl));
+                            }
```

Comment:
As per the other comment, all it takes is generics (https://godbolt.org/z/W8o3Gxhdo):
```rust
#[repr(simd)]
pub struct Vec4<T>([T; 4]);

pub fn ones() -> Vec4<NonZeroU8> {
    Vec4([NonZeroU8::new(1).unwrap(); 4])
}
```